tidy3d/components/time_modulation.py

```diff
@@ -0,0 +1,207 @@
+
+    """
+
+    @cached_property
+    @abstractmethod
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+
+
+class SpaceModulation(AbstractSpaceModulation):
+    """The modulation profile in space with user-supplied spatial distribution of
+    amplitude and phase.
+
+    Note
+    ----
+    .. math::
+
+        amp\\_space(r) = amplitude(r) \\cdot \\e^{i \\cdot phase(r)}
+
+    The full space-time modulation is,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+    Example
+    -------
+    >>> Nx, Ny, Nz = 10, 9, 8
+    >>> X = np.linspace(-1, 1, Nx)
+    >>> Y = np.linspace(-1, 1, Ny)
+    >>> Z = np.linspace(-1, 1, Nz)
+    >>> coords = dict(x=X, y=Y, z=Z)
+    >>> amp= SpatialDataArray(np.random.random((Nx, Ny, Nz)), coords=coords)
+    >>> phase= SpatialDataArray(np.random.random((Nx, Ny, Nz)), coords=coords)
+    >>> space = SpaceModulation(amplitude=amp, phase=phase)
+    """
+
+    amplitude: Union[float, SpatialDataArray] = pd.Field(
+        1,
+        title="Amplitude of modulation in space",
+        description="Amplitude of modulation that can vary spatially.",
+    )
+
+    phase: Union[float, SpatialDataArray] = pd.Field(
+        0,
+        title="Phase of modulation in space",
+        description="Phase of modulation that can vary spatially.",
+    )
+
+    interp_method: InterpMethod = pd.Field(
+        "nearest",
+        title="Interpolation method",
+        description="Interpolation method to obtain space modulation values "
+        "that are not supplied at the Yee grids.",
+    )
+
+    @pd.validator("amplitude", always=True)
+    def _real_amplitude(cls, val):
+        """Assert that the amplitude is real."""
+        if np.iscomplexobj(val):
+            raise ValidationError("'amplitude' must be real.")
+        return val
+
+    @pd.validator("phase", always=True)
+    def _real_phase(cls, val):
+        """Assert that the phase is real."""
+        if np.iscomplexobj(val):
+            raise ValidationError("'phase' must be real.")
+        return val
+
+    @cached_property
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+        return np.max(np.array(self.amplitude))
+
+
+SpaceModulationType = Union[SpaceModulation]
+
+
+class SpaceTimeModulation(Tidy3dBaseModel):
+    """Space-time modulation applied to a medium, adding
+    on top of the time-independent permittivity of the medium.
+
+
+    Note
+    ----
+    The space-time modulation must be separable in space and time.
+    e.g. when applied to permittivity,
+
+    .. math::
+
+        \\delta \\epsilon(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+    """
+
+    space_modulation: SpaceModulationType = pd.Field(
+        SpaceModulation(),
+        title="Space modulation",
+        description="Space modulation from the separable space-time modulation.",
+        # discriminator=TYPE_TAG_STR,
```

Comment:
probably best to have these for now, unless it causes errors? (just for future)

Reply:
Yeah, it errors when there is only one in union

Reply:
oh ok, good to know.